storage: work around can't-swap-leaseholder

tbg opened the pull request:
As of #40284, the replicate queue was issuing swaps (atomic add+remove)
during rebalancing. TestInitialPartitioning helpfully points out (once you
flip atomic rebalancing on) that when the replication factor is one, there
is no way to perform such an atomic swap because it will necessarily have
to remove the leaseholder.

To work around this restriction (which, by the way, we dislike - see
\#40333), fall back to just adding a replica in this case without also
removing one. In the next scanner cycle (which should happen immediately
since we requeue the range) the range will be over-replicated and hopefully
the lease will be transferred over and then the original leaseholder
removed. I would be very doubtful that this all works, but it is how things
worked until #40284, so this PR really just falls back to the previous
behavior in cases where we can't do better.

Release note: None

tbg committed:
storage: remove errant log message

There may be nothing to roll back, so don't log unconditionally.

Release note: None

tbg committed:
storage: make an error benign

This was showing up a lot in TestInitialPartitioning. If we're trying to
remove something but nothing needs to be removed, that seems OK (though
there is some question of why we're hitting this regularly).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/40363)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40363#-:-LnqcJB40iGEqSDW7uSA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/replicate_queue.go, line 515 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40363#-LnqbfSi6zamHwwThmdT:-LnqbfSi6zamHwwThmdU:br5g9yz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f10689362a2ec6e23808defa3c25c69f32169598/pkg/storage/replicate_queue.go#L515)):*
> ```Go
> // likely to be the leaseholder), then this removal would fail. Instead, this
> // method will attempt to transfer the lease away, and returns true to indicate
> // to the caller that it should not pursue the current replication change further.
> ```

"because it is no longer the leaseholder"
___
*[pkg/storage/replicate_queue.go, line 739 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40363#-Lnqc0K2B7hd7Cv1eoyo:-Lnqc0K2B7hd7Cv1eoyp:b-c46m30) ([raw file](https://github.com/cockroachdb/cockroach/blob/f10689362a2ec6e23808defa3c25c69f32169598/pkg/storage/replicate_queue.go#L739)):*
> ```Go
> 				// only, which should succeed, and the next time we touch this
> 				// range, we will have one more replica and hopefully it will
> 				// take the lease and remove the current leaseholder.
> ```

I'm surprised that this case doesn't hit an error when it calls `maybeTransferLeaseAway`. Could you mention what we expect to happen when you call that?


<!-- Sent from Reviewable.io -->

tbg committed:
storage: work around can't-swap-leaseholder

As of #40284, the replicate queue was issuing swaps (atomic add+remove)
during rebalancing.
TestInitialPartitioning helpfully points out (once you flip atomic
rebalancing on) that when the replication factor is one, there is
no way to perform such an atomic swap because it will necessarily
have to remove the leaseholder.

To work around this restriction (which, by the way, we dislike - see
\#40333), fall back to just adding a replica in this case without also
removing one. In the next scanner cycle (which should happen immediately
since we requeue the range) the range will be over-replicated and
hopefully the lease will be transferred over and then the original
leaseholder removed. I would be very doubtful that this all works,
but it is how things worked until #40284, so this PR really just
falls back to the previous behavior in cases where we can't do
better.

Release note: None

tbg reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40363#-:-LnrTMxbBhvvIREtQjlu:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

tbg commented:
bors r=nvanbenschoten
